src/webview/html/messageAsHtml.js

```diff
@@ -72,13 +77,87 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const widgets = message.submessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      ...submessage,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const poll_widgets = widgets.filter(widget => widget.content.widget_type === 'poll');
+  // "events" are submessages other than the main widget submessage - for
+  // instance, `vote`, `new_option`, and similar submessages. These are not
+  // documented, look at the PollData code if you're interested in what the
+  // possibilities are.
+  const events = widgets.filter(widget => 'type' in widget.content);
+
+  if (poll_widgets.length !== 1) {
+    return template`
 $!${message.content}
-<div class="special-message"
- ><p>Interactive message</p
- ><p>To use, open on web or desktop</p
-></div>
+<div class="special-message">
+  <p>Interactive message</p>
+  <p>To use, open on web or desktop</p>
+</div>
```

Comment:
This adds whitespace which can potentially affect how the HTML gets rendered. In general I prefer to avoid doing that, because when it does turn out to matter it isn't fun tracking down why there's a stray space somewhere.

(In any case this would be an independent change.)

Reply:
Removed this change.

Reply:
Interesting; I'll keep this in mind about whitespace. I looked at the old code and thought it was super weird; now I understand why we'd want it that way 🙂